src/libstd/collections/hash/map.rs

```diff
@@ -1370,6 +1370,37 @@ impl<K, V, S> Eq for HashMap<K, V, S>
 {
 }
 
+#[unstable(feature = "hashmap_hash", issue = "0")]
+impl<K, V, S> Hash for HashMap<K, V, S>
+    where K: Eq + Hash,
```

Comment:
Is the `Eq` bound necessary here?

Reply:
Yes, it is necessary for `self.iter()` - you can't call that otherwise. You could use `self.table.iter()` which would not require `Eq`. So `self.iter()` could be changed to not require `Eq`. But I think it is right to require `K: Eq` since you shouldn't be able to do anything useful with a `HashMap<K, V, S>` unless `K: Eq`.